pkg/daemon/certificate_writer.go

```diff
@@ -101,92 +114,317 @@ func (dn *Daemon) syncControllerConfigHandler(key string) error {
 	// Write the latest cert to disk, if the controllerconfig resourceVersion has updated
 	// Also annotate the latest config we've seen, so as to not write unnecessarily
 	currentNodeControllerConfigResource := dn.node.Annotations[constants.ControllerConfigResourceVersionKey]
+	var cmErr error
+	var data []byte
+	kubeConfigDiff := false
+	allCertsThere := true
+	dontRestartKubelet := false
 
-	if currentNodeControllerConfigResource != controllerConfig.ObjectMeta.ResourceVersion {
+	if currentNodeControllerConfigResource != controllerConfig.ObjectMeta.ResourceVersion || controllerConfig.Annotations[ctrlcommon.ServiceCARotateAnnotation] == ctrlcommon.ServiceCARotateTrue {
 		pathToData := make(map[string][]byte)
-
 		kubeAPIServerServingCABytes := controllerConfig.Spec.KubeAPIServerServingCAData
 		cloudCA := controllerConfig.Spec.CloudProviderCAData
 		pathToData[caBundleFilePath] = kubeAPIServerServingCABytes
 		pathToData[cloudCABundleFilePath] = cloudCA
+		var err error
+		var cm *corev1.ConfigMap
+		var FullCA []string
 
-		for bundle, data := range pathToData {
-			if !strings.HasSuffix(string(data), "\n") {
-				bString := string(data) + "\n"
-				data = []byte(bString)
-			}
-			if Finfo, err := os.Stat(bundle); err == nil {
-				var mode os.FileMode
-				Dinfo, err := os.Stat(filepath.Dir(bundle))
+		if controllerConfig.Annotations[ctrlcommon.ServiceCARotateAnnotation] == ctrlcommon.ServiceCARotateTrue && dn.node.Annotations[constants.ControllerConfigSyncServerCA] != controllerConfig.Annotations[ctrlcommon.ServiceCARotateAnnotation] {
+			cm, cmErr = dn.kubeClient.CoreV1().ConfigMaps("openshift-machine-config-operator").Get(context.TODO(), "kubeconfig-data", v1.GetOptions{})
+			if cmErr != nil {
+				klog.Errorf("Error retrieving kubeconfig-data. %v", cmErr)
+			} else {
+				data, err = cmToData(cm, "ca-bundle.crt")
 				if err != nil {
-					mode = defaultDirectoryPermissions
-				} else {
-					mode = Dinfo.Mode()
+					klog.Errorf("kubeconfig-data ConfigMap not populated yet. %v", err)
+				} else if data != nil {
+					currentKC := clientcmdv1.Config{}
+					kcBytes, err := os.ReadFile(kubeConfigPath)
+					if err != nil {
+						return err
+					}
+					if kcBytes != nil {
+						err = yaml.Unmarshal(kcBytes, &currentKC)
+						if err != nil {
+							return fmt.Errorf("could not unmarshal kubeconfig into struct. Data: %s, Error: %v", string(kcBytes), err)
+						}
+						kubeConfigDiff = !bytes.Equal(bytes.TrimSpace(currentKC.Clusters[0].Cluster.CertificateAuthorityData), bytes.TrimSpace(data))
+						logSystem("Current data %s and New Data %s", string(currentKC.Clusters[0].Cluster.CertificateAuthorityData), string(data))
```

Comment:
should probably remove this log line, I don't think we should have it outside of debugging purposes